### PR TITLE
Hibernate 5.2 support by reflection

### PR DESCRIPTION
**Pull request by mshima**
Since hibernate 5.2 breaks abi compatibility, this commit add a workaround using reflection.
This PR fixes at least SERIALIZE_IDENTIFIER_FOR_LAZY_NOT_LOADED_OBJECTS feature on hibernate 5.
Build works against hibernate 5.2.4.Final too.

### CODE REVIEW

File: hibernate5/src/main/java/com/fasterxml/jackson/datatype/hibernate5/HibernateProxySerializer.java [hunk 1]
```diff
@@ -182,8 +191,28 @@ protected Object findProxied(HibernateProxy proxy)
                 if (_mapping != null) {
                     idName = _mapping.getIdentifierPropertyName(init.getEntityName());
                 } else {
-                    final SessionImplementor session = init.getSession();
-                    if (session != null) {
+                    // Detect abi compatibility if not set yet
+                    if(hibernateAbiIncompatible == null){
```
Comment (cowtowncoder) [inline hibernate5/src/main/java/com/fasterxml/jackson/datatype/hibernate5/HibernateProxySerializer.java]: I think this should be done just once from constructor or even static block: calling it every time serialization occurs is going to be quite inefficient.

File: hibernate5/src/main/java/com/fasterxml/jackson/datatype/hibernate5/HibernateProxySerializer.java [hunk 2]
```diff
@@ -182,8 +191,28 @@ protected Object findProxied(HibernateProxy proxy)
...
+                        } catch (NoSuchMethodError e) {
+                            hibernateAbiIncompatible = true;
+                        }
+                    }
+                    final Object sessionObj;
+                    if(!hibernateAbiIncompatible){
+                        sessionObj = init.getSession();
+                    }else{
+                        try {
+                            Method method = init.getClass().getMethod("getSession");
```
Comment (cowtowncoder) [inline hibernate5/src/main/java/com/fasterxml/jackson/datatype/hibernate5/HibernateProxySerializer.java]: Similarly here it should be possible do this just once, locating method is not free (reflection API is somewhat inefficient wrt lookups); should be doable along with earlier detection.

**Comment by cowtowncoder** [comment]
First of all, thank you for providing this patch. I assume you have tried it with both 5.2 and earlier versions? (5.0 / 5.1).

I added some minor notes on doing expensive ABI checks upfront, instead of each and every time values are serialized.

Other than that, I think it's probably good; I don't know Hibernate well enough to really judge it but as long as it works around the issue that's improvement.

File: hibernate5/src/main/java/com/fasterxml/jackson/datatype/hibernate5/HibernateProxySerializer.java (hunk 2 above)
Comment (mshima) [inline hibernate5/src/main/java/com/fasterxml/jackson/datatype/hibernate5/HibernateProxySerializer.java]: It's possible to keep the method object on a variable. The getMethod call can be skipped after the first call.

**Comment by mshima** [comment]
We need to know what the version the module is built with.
Then we need to detect what version is running ( can be done with Version class or detecting if some class exists).
After that we can verify if the abi is compatible without executing the method getSession and getting the NoSuchMethodError.
The first step is the problem.

**Comment by mshima** [comment]
Reworked the PR.
- Always execute using reflection instead of detecting incompatibility.
- Create an inner class and put the logic like the existing ProxyReader.

**Comment by mshima** [comment]
Continuation of hibernate 5.2 support:
- Fix for https://github.com/FasterXML/jackson-datatype-hibernate/issues/99 is https://github.com/mshima/jackson-datatype-hibernate/commit/3d7fb5d5456cfdc2b90d586e729f25ad20e91d8f
- Test cases against hibernate 5.2 is https://github.com/mshima/jackson-datatype-hibernate/commit/75ae4c8989a38b5c8093980628f7ada6015ea773


After all I started wondering if this is the way to go: https://github.com/mshima/jackson-datatype-hibernate/commit/120d5ef7c551ad90598e923ad48b2e92c0527c5f
- Create a hibernate52 module using hibernate5 pom.xml.
- Change the source dir to hibernate5 module.
- Minor fixes.
- Change the build jvm java8 for hibernate 5.2 compatibility.
- Throw everything else away.

**Comment by z0mb1ek** [comment]
when can we use it?have you snapshot repos?thx

**Comment by cowtowncoder** [comment]
@mshima On version detection; is this wrt Hibernate version, or Jackson version(s)? Latter is possible by using generated `PackageVersion` of Hibernate modules as well as databind -- I fixed this for Hibernate module since it was still using obsolete mechanism.

@z0mb1ek eventually once this gets merged it may be used, I can push new snapshot manually, no automated build yet exists. Local build should be easy enough as well.

**Comment by mshima** [comment]
_@cowtowncoder Hibernate version. The problem is ABI compatibility not API.
If jackson-datatype-hibernate5 is built with hibernate 5.1 (like now), at runtime it will work with hibernate 5.0 and 5.1, but will fail with 5.2.
If jackson-datatype-hibernate5 is built with hibernate 5.2, at runtime it will work with hibernate 5.2, but will fail with 5.0 and 5.1.

I don't know which approach is better:
- Reflection (The code became complicated)
or
- Another module just with different build dependency (the build jvm must be changed to java8)

**Comment by cowtowncoder** [comment]
@z0mb1ek yes, once this gets merged I can push a snapshot.

**Comment by cowtowncoder** [comment]
@mshima Ah. I was afraid this might be the case.

In a way a separate sub-module would be safest way, and we already have 3. But for maximal backwards compatibility, it would be necessary to have new version like `hibernate5_2` and leave existing `hibernate5` for 5.0 and 5.1.
I am not really happy with this approach since there is a lot of code duplication and no good way to solve this.

If at all possible, I would prefer reflection based approach, but I don't know how easy it would be to do this so that single build (against 5.2 I presume) would be possible. Further use of reflection should be coarse-grained -- something like loading appropriate component(s) upon construction. But then again perhaps more dynamic approach would be needed... 

So I guess I'd be ok with "new `hibernate5_2` module" approach: cut'n pasting `hibernate5` and making modifications from there.

**Comment by mshima** [comment]
Pushed the remaining fixes.

Created https://github.com/FasterXML/jackson-datatype-hibernate/pull/101 with "new hibernate5_2 module" but using same source dir from hibernate5 module, without copy and paste.

File: hibernate5_2-test/pom.xml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      <activation>
+        <jdk>1.8</jdk>
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <configuration>
+              <dependenciesToScan>
```
Comment (cowtowncoder) [inline hibernate5_2-test/pom.xml]: Could you add a comment about what this does? I'd assume regular dependency mechanism won't do for some reason?

File: hibernate5/pom.xml [hunk 4]
```diff
@@ -97,6 +97,18 @@ Hibernate (http://hibernate.org) version 5.x data types.
   <build>
     <plugins>
       <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
```
Comment (cowtowncoder) [inline hibernate5/pom.xml]: Is this added to produce a separate test-jar, to be used by 5.2 test package?

**Comment by cowtowncoder** [comment]
@mshima Apologies for slow followup. Looks good in general; added couple of small questions to understand some of the details. One minor thing that would be good (if not done already) is to prevent packaging/export of 5.2-test jar.

File: hibernate5/pom.xml (hunk 4 above)
Comment (mshima) [inline hibernate5/pom.xml]: yes

File: hibernate5_2-test/pom.xml (hunk 3 above)
Comment (mshima) [inline hibernate5_2-test/pom.xml]: Done

**Comment by z0mb1ek** [comment]
when i can get it from repos?

**Comment by cowtowncoder** [comment]
@mshima Ok just found out one problem: `mvn deploy` fails due to missing artifact for hibernate5_2-test. Will try to see how to prevent deploy plug-in from croaking.

**Comment by cowtowncoder** [comment]
@z0mb1ek whenever we cut releases there will be official new versions. This won't happen for a while with 2.8.6 both because number of fixes is slowing down, and because I'll be off for vacation until january.
But I did just push 2.8.6-SNAPSHOT for sonatype snapshot repo which may work.

**Comment by z0mb1ek** [comment]
@cowtowncoder big thanks!!!

**Comment by cowtowncoder** [comment]
@mshima Just had to add `skip` for deploy as well, taken care of now. Also applied install plug-in version update to parent poms for jackson 2.9.

**Comment by mshima** [comment]
@cowtowncoder great.

**Comment by mshima** [comment]
@cowtowncoder https://github.com/FasterXML/jackson-datatype-hibernate/issues/99 should be fixed too.

**Pull request merged**